src/bun.js/node/types.zig

```diff
@@ -1840,548 +1849,2904 @@ pub const Emitter = struct {
+        defer path.deinit();
 
-            if (u16_slice.len > 3)
-                buf[3] = u16_slice[3];
+        const out = if (!isWindows)
+            dirnamePosix(path.slice())
+        else
+            dirnameWindows(path.slice());
+        return JSC.ZigString.init(out).withEncoding().toValueGC(globalObject);
+    }
+    
+    /// Based on Node v21.6.1 path.posix.extname:
+    /// https://github.com/nodejs/node/blob/6ae20aa63de78294b18d5015481485b7cd8fbb60/lib/path.js#L1278
+    pub fn extnamePosixT(comptime T: type, path: []const T) []const T {
+        // validateString of `path` is performed in fn pub extname.
+        const len = path.len;
+        // Exit early for easier number type use.
+        if (len == 0) {
+            return comptime L(T, "");
+        }
+        // We use an optional value instead of -1, as in Node code, for easier number type use.
+        var startDot: ?usize = null;
+        var startPart: usize = 0;
+        // We use an optional value instead of -1, as in Node code, for easier number type use.
+        var end: ?usize = null;
+        var matchedSlash: bool = true;
+        // Track the state of characters (if any) we see before our first dot and
+        // after any path separator we find
+        
+        // We use an optional value instead of -1, as in Node code, for easier number type use.
+        var preDotState: ?usize = 0;
+        
+        var i_i64 = @as(i64, @intCast(len - 1));
+        while (i_i64 > -1) : (i_i64 -= 1) {
+            const i = @as(usize, @intCast(i_i64));
+            const byte = path[i];
+            if (byte == CHAR_FORWARD_SLASH) {
+                // If we reached a path separator that was not part of a set of path
+                // separators at the end of the string, stop now
+                if (!matchedSlash) {
+                    startPart = i + 1;
+                    break;
+                }
+                continue;
+            }
+            
+            if (end == null) {
+                // We saw the first non-path separator, mark this as the end of our
+                // extension
+                matchedSlash = false;
+                end = i + 1;
+            }
+            
+            if (byte == CHAR_DOT) {
+                // If this is our first dot, mark it as the start of our extension
+                if (startDot == null) {
+                    startDot = i;
+                } else if (preDotState != null and preDotState.? != 1) {
+                    preDotState = 1;
+                }
+            } else if (startDot != null) {
+                // We saw a non-dot and non-path separator before our dot, so we should
+                // have a good chance at having a non-empty extension
+                preDotState = null;
+            }
+        }
 
-            return std.fs.path.isAbsoluteWindowsWTF16(buf[0..@min(u16_slice.len, buf.len)]);
+        const _end = if (end)|__|__ else 0;
```

Comment:
not important but for this pattern you can do `... = end orelse 0;`

Reply:
Tweaked, its unwrapped an optional value, but I made it more descriptive.